File: docs/src/test/scala/docs/home/persistence/CouchbaseReadSideQuery.scala
/*
 * Copyright (C) 2009-2018 Lightbend Inc. <https://www.lightbend.com>
 */
package docs.home.persistence

// #imports
import akka.NotUsed
import akka.stream.alpakka.couchbase.scaladsl.CouchbaseSession
import com.lightbend.lagom.scaladsl.api.{Service, ServiceCall}
import com.lightbend.lagom.scaladsl.persistence.ReadSide
import com.lightbend.lagom.scaladsl.persistence.couchbase.CouchbaseReadSide
import docs.home.persistence.CouchbaseReadSideProcessorTwo.HelloEventProcessor
import play.api.libs.json.{Format, Json}

import scala.collection.JavaConverters._
import scala.concurrent.ExecutionContext

// #imports

object CouchbaseReadSideQuery {

  trait GreetingService extends Service {
    def userGreetings(): ServiceCall[NotUsed, List[UserGreeting]]

    override final def descriptor = {
      import Service._
      named("hello")
        .withCalls(
          pathCall("/api/user-greetings/", userGreetings _)
        )
        .withAutoAcl(true)
    }
  }

  case class UserGreeting(name: String, message: String)
  object UserGreeting {
    implicit val format: Format[UserGreeting] = Json.format[UserGreeting]
  }

  // #service-impl
  // FIXME both ReadSide and CouchbaseReadSide what is wrong here?
  class GreetingServiceImpl(readSide: CouchbaseReadSide with ReadSide, session: CouchbaseSession)(
      implicit ec: ExecutionContext
  ) extends GreetingService {
    readSide.register[HelloEvent](new HelloEventProcessor(readSide))
    override def userGreetings() =
      ServiceCall { request =>
        session.get("users-actual-greetings").map {
          case Some(jsonDoc) =>
            val json = jsonDoc.content()
            json.getNames().asScala.map(name => UserGreeting(name, json.getString(name))).toList
          case None => List.empty[UserGreeting]
        }
      }
  }
  // #service-impl

  //#register-event-processor
  ???
  //#register-event-processor
}

File: docs/src/main/paradox/lagom/java/code/docs/home/persistence/CouchbaseReadSideQuery.java
package docs.home.persistence;

//#imports
import akka.NotUsed;
import akka.stream.alpakka.couchbase.javadsl.CouchbaseSession;
import com.couchbase.client.java.document.json.JsonObject;
import com.lightbend.lagom.javadsl.api.ServiceCall;

import java.util.Collections;
import java.util.List;
import java.util.stream.Collectors;
import javax.inject.Inject;
//#imports

public interface CouchbaseReadSideQuery {

  interface GreetingService {
    ServiceCall<NotUsed, List<UserGreeting>> userGreetings();
  }

  //#service-impl
  public class GreetingServiceImpl implements GreetingService {

    private final CouchbaseSession couchbaseSession;

    @Inject
    public GreetingServiceImpl(CouchbaseSession couchbaseSession) {
      this.couchbaseSession = couchbaseSession;
    }

    @Override
    public ServiceCall<NotUsed, List<UserGreeting>> userGreetings() {
      return request -> couchbaseSession.get("users-actual-greetings")
          .thenApply(docOpt -> {
            if (docOpt.isPresent()) {
              JsonObject content = docOpt.get().content();
              return content.getNames().stream().map(
                  name -> new UserGreeting(name, content.getString(name))
              ).collect(Collectors.toList());
            } else {
              return Collections.emptyList();
            }
          });
    }
  }
  //#service-impl

  public class GreetingServiceImpl2 implements GreetingService {

    //#register-event-processor
    @Inject
    public GreetingServiceImpl(ReadSide readSide) {

      readSide.register(HelloEventProcessor.class);
    }
    //#register-event-processor

    @Override
    public ServiceCall<NotUsed, List<UserGreeting>> userGreetings() {
      return request -> couchbaseSession.get("users-actual-greetings")
          .thenApply(docOpt -> {
            if (docOpt.isPresent()) {
              JsonObject content = docOpt.get().content();
              return content.getNames().stream().map(
                  name -> new UserGreeting(name, content.getString(name))
              ).collect(Collectors.toList());
            } else {
              return Collections.emptyList();
            }
          });
    }
  }

}
